### PR TITLE
Automatically select clustering tool

### PR DESCRIPTION
**Pull request by mzueva**
Automatically switch to linear clustering for datasets with more than 1M rows

<h3>Greptile Summary</h3>

This PR auto-selects the clustering algorithm based on dataset size: when the selected dataset exceeds 1 million rows, the UI switches `clusteringTool` to `easy-linclust` (linear-time); below the threshold it keeps `easy-cluster`. It also bumps several `@platforma-sdk` packages to 1.79.14.

**Key touched terms:**
- **`LARGE_DATASET_ROW_THRESHOLD`** — New exported constant (`1_000_000`). Defines the row-count boundary that triggers auto-selection of `easy-linclust` over `easy-cluster` in the UI watcher.
- **`datasetMatchers`** — Refactored in this PR from an inline literal to a module-private named constant, reused by the `datasetOptions` output and potentially by future outputs.
- **`datasetSize`** (output) — New `BlockModelV3` output returning `number | undefined`. Calls `getNumberOfRows(col.data)` on the selected column; returns `undefined` when no dataset is selected, the column isn't resolved yet, or the backing storage is non-parquet.
- **`clusteringTool`** (`BlockData` field) — Pre-existing `\"easy-cluster\" | \"easy-linclust\"` field. Now reactively auto-set by the new watcher in `MainPage.vue` based on `datasetSize`, while still being editable manually in Advanced Settings.
- **`getNumberOfRows`** — Newly imported SDK function. Reads parquet chunk statistics from column resource metadata without downloading blob data; the mechanism that makes `datasetSize` cheap to compute.

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

The change is small and well-scoped; the new watcher writes a bounded enum value, dependency bumps are in-lockstep, and the fallback (undefined size → no change) is correctly handled.

The auto-selection logic is sound, but there is a narrow window where a manual tool choice made while datasetSize is still resolving will be silently overridden once the value arrives. The JSDoc for LARGE_DATASET_ROW_THRESHOLD also references a non-existent symbol. Neither issue affects correctness for the common path, but the silent override could surprise users in slow-loading environments.

The watcher block in ui/src/pages/MainPage.vue (lines 92-99) and the JSDoc on LARGE_DATASET_ROW_THRESHOLD in model/src/index.ts are the only areas worth a second look before merging.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| model/src/index.ts | Adds LARGE_DATASET_ROW_THRESHOLD, extracts datasetMatchers constant, and introduces the datasetSize output; one stale JSDoc reference to a non-existent datasetSizeByRef symbol. |
| ui/src/pages/MainPage.vue | Adds a Vue watcher on datasetSize that auto-selects easy-linclust for large datasets; well-commented with a potential edge case where a manual tool choice made during data loading can be silently overridden. |
| pnpm-workspace.yaml | Bumps @platforma-sdk/model, @platforma-sdk/ui-vue, @platforma-sdk/test to 1.79.14 and @platforma-sdk/workflow-tengo to 6.6.3; straightforward dependency updates. |
| pnpm-lock.yaml | Lock file updated to reflect dependency version bumps; no manual edits, consistent with workspace catalog changes. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant User
    participant MainPage as MainPage.vue
    participant BlockModel as BlockModel (datasetSize output)
    participant SDK as @platforma-sdk/model getNumberOfRows

    User->>MainPage: Selects dataset (PlDropdownRef)
    MainPage->>BlockModel: "data.datasetRef = inputRef (setInput)"
    BlockModel->>SDK: getNumberOfRows(col.data)
    SDK-->>BlockModel: "number | undefined (parquet chunk stats)"
    BlockModel-->>MainPage: outputs.datasetSize (reactive)
    MainPage->>MainPage: watch(datasetSize) fires
    alt "size === undefined (loading / non-parquet)"
        MainPage->>MainPage: leave clusteringTool unchanged
    else "size > LARGE_DATASET_ROW_THRESHOLD (1 000 000)"
        MainPage->>BlockModel: "data.clusteringTool = easy-linclust"
    else "size <= threshold"
        MainPage->>BlockModel: "data.clusteringTool = easy-cluster"
    end
    User->>MainPage: Manually overrides tool in Advanced Settings
    Note over MainPage,BlockModel: datasetSize unchanged - watcher silent - choice preserved
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant User
    participant MainPage as MainPage.vue
    participant BlockModel as BlockModel (datasetSize output)
    participant SDK as @platforma-sdk/model getNumberOfRows

    User->>MainPage: Selects dataset (PlDropdownRef)
    MainPage->>BlockModel: "data.datasetRef = inputRef (setInput)"
    BlockModel->>SDK: getNumberOfRows(col.data)
    SDK-->>BlockModel: "number | undefined (parquet chunk stats)"
    BlockModel-->>MainPage: outputs.datasetSize (reactive)
    MainPage->>MainPage: watch(datasetSize) fires
    alt "size === undefined (loading / non-parquet)"
        MainPage->>MainPage: leave clusteringTool unchanged
    else "size > LARGE_DATASET_ROW_THRESHOLD (1 000 000)"
        MainPage->>BlockModel: "data.clusteringTool = easy-linclust"
    else "size <= threshold"
        MainPage->>BlockModel: "data.clusteringTool = easy-cluster"
    end
    User->>MainPage: Manually overrides tool in Advanced Settings
    Note over MainPage,BlockModel: datasetSize unchanged - watcher silent - choice preserved
```

</a>
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Amodel%2Fsrc%2Findex.ts%3A40-41%0A**Stale%20symbol%20in%20JSDoc%20comment**%0A%0AThe%20JSDoc%20says%20%60see%20%5C%60datasetSizeByRef%5C%60%20and%20%5C%60setInput%5C%60%60%20but%20%60datasetSizeByRef%60%20does%20not%20exist%20anywhere%20in%20the%20codebase.%20The%20actual%20implementation%20uses%20the%20%60datasetSize%60%20output%20%28declared%20three%20lines%20below%29%20and%20the%20%60watch%60%20in%20%60MainPage.vue%60.%20A%20developer%20following%20this%20reference%20will%20hit%20a%20dead%20end.%0A%0A%23%23%23%20Issue%202%20of%203%0Amodel%2Fsrc%2Findex.ts%3A40%0AThe%20JSDoc%20references%20%60datasetSizeByRef%60%20which%20doesn't%20exist%20%E2%80%94%20the%20actual%20output%20is%20named%20%60datasetSize%60.%0A%0A%60%60%60suggestion%0A%20*%20on%20dataset%20selection%20in%20the%20UI%20%E2%80%94%20see%20%60datasetSize%60%20output%20and%20the%20%60watch%60%20in%20%60MainPage.vue%60.%0A%60%60%60%0A%0A%23%23%23%20Issue%203%20of%203%0Aui%2Fsrc%2Fpages%2FMainPage.vue%3A92-99%0A**Auto-selection%20can%20override%20a%20manual%20choice%20made%20while%20data%20is%20still%20loading**%0A%0AThe%20comment%20says%20%22a%20later%20manual%20choice%20in%20Advanced%20Settings%20is%20preserved%22%20%E2%80%94%20this%20is%20true%20once%20%60datasetSize%60%20has%20settled%2C%20because%20the%20watcher%20only%20fires%20on%20changes.%20However%2C%20if%20a%20user%20opens%20the%20settings%20panel%20and%20manually%20picks%20a%20tool%20before%20%60datasetSize%60%20resolves%20from%20%60undefined%60%20to%20a%20concrete%20number%20%28e.g.%2C%20first%20open%20of%20a%20block%20with%20a%20freshly-added%20dataset%29%2C%20the%20watcher%20will%20still%20fire%20when%20the%20value%20arrives%20and%20will%20overwrite%20that%20choice.%20The%20guard%20%60if%20%28size%20%3D%3D%3D%20undefined%29%20return%3B%60%20prevents%20writing%20on%20%60undefined%60%2C%20but%20it%20does%20not%20prevent%20overwriting%20a%20manual%20selection%20made%20in%20the%20window%20between%20dataset%20selection%20and%20data%20resolution.%0A%0A&repo=platforma-open%2Fclonotype-clustering&pr=127&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 3 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 3
model/src/index.ts:40-41
**Stale symbol in JSDoc comment**

The JSDoc says `see \`datasetSizeByRef\` and \`setInput\`` but `datasetSizeByRef` does not exist anywhere in the codebase. The actual implementation uses the `datasetSize` output (declared three lines below) and the `watch` in `MainPage.vue`. A developer following this reference will hit a dead end.

### Issue 2 of 3
model/src/index.ts:40
The JSDoc references `datasetSizeByRef` which doesn't exist — the actual output is named `datasetSize`.

```suggestion
 * on dataset selection in the UI — see `datasetSize` output and the `watch` in `MainPage.vue`.
```

### Issue 3 of 3
ui/src/pages/MainPage.vue:92-99
**Auto-selection can override a manual choice made while data is still loading**

The comment says "a later manual choice in Advanced Settings is preserved" — this is true once `datasetSize` has settled, because the watcher only fires on changes. However, if a user opens the settings panel and manually picks a tool before `datasetSize` resolves from `undefined` to a concrete number (e.g., first open of a block with a freshly-added dataset), the watcher will still fire when the value arrives and will overwrite that choice. The guard `if (size === undefined) return;` prevents writing on `undefined`, but it does not prevent overwriting a manual selection made in the window between dataset selection and data resolution.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["remove obvious comments"](https://github.com/platforma-open/clonotype-clustering/commit/ff58c40fbde1eaaa3b1809e0e25eeae952626d47) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38626521)</sub>

> Greptile also left **2 inline comments** on this PR.

**Context used:**

- Context used - Terms is a types in codebase. Provide the list of ... ([source](https://app.greptile.com/review/custom-context?memory=instruction-0))

### CODE REVIEW

File: ui/src/pages/MainPage.vue [hunk 1]
```diff
@@ -77,6 +78,26 @@ function setInput(inputRef?: PlRef) {
...
+// a later manual choice in Advanced Settings is preserved. Size unknown
+// (non-parquet / not yet ready) -> leave the current choice untouched.
+watch(
+  () => app.model.outputs.datasetSize,
+  (size) => {
+    if (size === undefined) return;
+    app.model.data.clusteringTool =
+      size > LARGE_DATASET_ROW_THRESHOLD ? "easy-linclust" : "easy-cluster";
+  },
+);
```
[the remark this answers is not in the document]
Comment (mzueva) [reply ui/src/pages/MainPage.vue:98]: fixed

File: model/src/index.ts [hunk 2]
```diff
@@ -33,6 +34,29 @@ export const clusteringToolOptions = [
   { label: "Easy Linclust", value: "easy-linclust" },
 ] as const;
 
+/**
+ * Above this many input rows, the cascaded easy-cluster algorithm becomes
+ * prohibitively slow; linclust (linear-time) is auto-selected instead. Applied
+ * on dataset selection in the UI — see `datasetSizeByRef` and `setInput`.
+ */
+export const LARGE_DATASET_ROW_THRESHOLD = 1_000_000;
```
[the remark this answers is not in the document]
Comment (mzueva) [reply model/src/index.ts:42]: fixed
